File: JavaScript/SearchingAlgorithm/1_Algorithms.js
// Searches One by One Element and time Complexity is => O(n)
const LinearSearch = (arr, elem) => {
  for (let i = 0; i < arr.length; i++) {
    if (elem === arr[i]) return i;
  }
  return -1;
};

//console.log(LinearSearch([2, 3, 5, 63, 3, 6, 7, 8], 633));

// (Only on Sorted Array) Searches on Log Scale Element and time Complexity is => O(n)
const BinarySearch = (arr, elem) => {
  if (elem < arr[0] || elem > arr[arr.length - 1]) return -1; //Since its an Sorted Array this can reduce time.
  let left = 0;
  let right = arr.length - 1;
  let middle = Math.floor((left + right) / 2);
  while (arr[middle] !== elem) {
    console.log(left, middle, right);
    if (left >= right) return -1;
    if (arr[middle] < elem) left = middle + 1;
    if (arr[middle] > elem) right = middle - 1;
    middle = Math.floor((left + right) / 2);
  }
  return middle;
};

//console.log(BinarySearch([1, 2, 3, 4, 5, 55, 56, 71, 75], 55));

//Basic SubString Search O(n^2)
const NaiveStringSearch = (longString, shortstring) => {
  let count = 0;
  for (let i = 0; i < longString.length; i++) {
    for (let j = 0; j < shortstring.length; j++) {
      console.log(shortstring[j], longString[i + j]);
      if (shortstring[j] !== longString[i + j]) break;
      if (j === shortstring.length - 1) count++;
    }
  }
  return count;
};

console.log(NaiveStringSearch("i feel so so Good", "so"));

File: Python/3_SearchingAlgorithms/1_Algorithms.py
import math

def LinearSearch(lst, elem):
	for i in range(len(lst)):
		if lst[i] == elem:
			return i
	return -1


# only sorted array
def BinarySearch(lst, elem):
	if elem < lst[0] or elem > lst[-1]: return -1
	left = 0
	right = len(lst)
	mid = math.floor((left + right)/2)

	while lst[mid] != elem:
		if left >= right: return -1

		if lst[mid] > elem:
			right = mid - 1
		elif lst[mid] < elem:
			left = mid +1

		mid = math.floor((left + right)/2)
	return mid

def NaiveStringSearch(long_str, short_str):
	long_len = len(long_str)
	short_len = len(short_str)

	count = 0

	for i in range(long_len - short_len + 1):
		if long_str[i: i+short_len] == short_str:
			count += 1
	return count 